Audio device controls for mobile native audio device selection

toger5 opened the pull request:
Fixes: https://github.com/element-hq/element-call/issues/3262
See: [controlledOutput](https://github.com/element-hq/element-call/blob/f69c75322f0e247fd42a9d2aa3e24d4b6bb959be/docs/controls.md)

robintown committed:
WIP

toger5 committed:
Add custom audio renderer to only render joined participants & add ios earpice workaround

fix left right to match chromium + safari
(firefox is swapped)

earpice as setting

Simpler code and documentation
The doc explains, what this class actually does and why it is so complicated.

Signed-off-by: Timo K <toger5@hotmail.de>

use only one audioContext, remove (non working) standby fallback

toger5 committed:
Add tests

toger5 committed:
use optional audio context and effect to initiate it + review

toger5 committed:
temp

toger5 committed:
add earpice mode

toger5 committed:
turn on url flag by default

toger5 committed:
apply mute from mobile controls

robintown commented on docs/controls.md:
```diff
@@ -1,7 +1,18 @@
+These functions must be used in conjunction with the `controlledOutput` URL parameter in order to have any effect.
+
+- `controls.setOutputDevices(devices: { id: string, name: string, forEarpiece?: boolean }[]): void` Sets the list of available audio outputs. `forEarpiece` is used on ios only.
+  It flags the device that should be used if the user selects earpice mode. This should be the main (stereo loudspeaker) of the device.
```
```suggestion
  It flags the device that should be used if the user selects earpiece mode. This should be the main stereo loudspeaker of the device.
```

robintown commented on src/UrlParams.ts:
```diff
@@ -281,6 +284,8 @@ export const getUrlParams = (
     allowIceFallback: parser.getFlagParam("allowIceFallback"),
     perParticipantE2EE: parser.getFlagParam("perParticipantE2EE"),
+    // TODO this should not default to true!
+    controlledOutput: parser.getFlagParam("controlledMediaDevices", true),
```
Leaving this comment here so we can't forget the TODO

toger5 replied:
We have: https://github.com/matrix-org/matrix-rust-sdk/pull/5049 now se I remove the wrong default value.

robintown commented on src/App.tsx:
```diff
@@ -72,26 +91,15 @@ export const App: FC = () => {
+                    ) : (
+                      <MediaDevicesProvider>{inner}</MediaDevicesProvider>
+                    )}
+                  </ProcessorProvider>
```
As I hinted at the other week I think we can do

```diff
@@ -67,19 +67,7 @@ export const App: FC = () => {
       .catch(logger.error);
   });
 
-  const inner = (
-    <Sentry.ErrorBoundary
-      fallback={(error) => <ErrorPage error={error} widget={widget} />}
-    >
-      <DisconnectedBanner />
-      <Routes>
-        <SentryRoute path="/" element={<HomePage />} />
-        <SentryRoute path="/login" element={<LoginPage />} />
-        <SentryRoute path="/register" element={<RegisterPage />} />
-        <SentryRoute path="*" element={<RoomPage />} />
-      </Routes>
-    </Sentry.ErrorBoundary>
-  );
+  const MediaDevicesProviderComponent = controlledOutput ? ControlledOutputMediaDevicesProvider : MediaDevicesProvider
 
   return (
     // eslint-disable-next-line @typescript-eslint/ban-ts-comment
@@ -92,13 +80,19 @@ export const App: FC = () => {
               <Suspense fallback={null}>
                 <ClientProvider>
                   <ProcessorProvider>
-                    {controlledOutput ? (
-                      <ControlledOutputMediaDevicesProvider>
-                        {inner}
-                      </ControlledOutputMediaDevicesProvider>
-                    ) : (
-                      <MediaDevicesProvider>{inner}</MediaDevicesProvider>
-                    )}
+                    <MediaDevicesProviderComponent>
+                      <Sentry.ErrorBoundary
+                        fallback={(error) => <ErrorPage error={error} widget={widget} />}
+                      >
+                        <DisconnectedBanner />
+                        <Routes>
+                          <SentryRoute path="/" element={<HomePage />} />
+                          <SentryRoute path="/login" element={<LoginPage />} />
+                          <SentryRoute path="/register" element={<RegisterPage />} />
+                          <SentryRoute path="*" element={<RoomPage />} />
+                        </Routes>
+                      </Sentry.ErrorBoundary>
+                    </MediaDevicesProviderComponent>
                   </ProcessorProvider>
                 </ClientProvider>
               </Suspense>
```

and then it becomes a bit cleaner as `inner` remains inline.

toger5 replied:
Do you think the solution with just one context is good as well?

robintown commented on src/room/InCallView.tsx:
```diff
@@ -222,7 +223,9 @@ export const InCallView: FC<InCallViewProps> = ({
-  const [muteAllAudio] = useSetting(muteAllAudioSetting);
+  const muteAllAudioControlled = useObservableEagerState(setOutputEnabled$);
+  const [muteAllAudioFromSetting] = useSetting(muteAllAudioSetting);
+  const muteAllAudio = muteAllAudioControlled || muteAllAudioFromSetting;
```
I would expect this `setOutputEnabled$` subject to be consumed by the CallViewModel, piped through `startWith` to give it an initial value, combined with the setting there, and exposed as a single state observable. (I'm actually surprised if this isn't crashing the InCallView component currently because `setOutputEnabled$`, being a subject, doesn't provide an initial value.)

Tangentially, I think you forgot a `!`:

```suggestion
  const muteAllAudioControlled = !useObservableEagerState(setOutputEnabled$);
  const [muteAllAudioFromSetting] = useSetting(muteAllAudioSetting);
  const muteAllAudio = muteAllAudioControlled || muteAllAudioFromSetting;
```

toger5 replied:
I am not sure this is possible since we need the mute info also in the groupCallView

toger5 replied:
moved it into its own file and fixed the `!` mistake.

robintown commented on src/room/GroupCallView.tsx:
```diff
@@ -104,8 +106,9 @@ export const GroupCallView: FC<Props> = ({
   );
-
-  const [muteAllAudio] = useSetting(muteAllAudioSetting);
+  const muteAllAudioControlled = useObservableEagerState(setOutputEnabled$);
+  const [muteAllAudioFromSetting] = useSetting(muteAllAudioSetting);
+  const muteAllAudio = muteAllAudioControlled || muteAllAudioFromSetting;
```
Oh gosh, I guess this usage of `setOutputEnabled$` in the GroupCallView shows that, if we turn this into a state observable, that state observable needs to live not on the CallViewModel, but rather something a bit more global than that. Since we don't yet have anything outside the InCallView refactored to use MVVM, I suppose accepting a bit of duplication or putting the `muteAllAudio` observable somewhere in the global/module scope would be fine with me.

toger5 replied:
oh I see.

toger5 replied:
I moved this into a file called `MuteAllAudioModel.ts` With a comment that this might converge into a `GroupCallViewModel.ts`.
If that is what we think is appropritate. otherwise we can move it into a place where it fits once we have such a place.

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -59,12 +60,26 @@ export interface MediaDevice {
+
+export interface MediaDevices extends Omit<InputDevices, "usingNames"> {
+  audioOutput: MediaDevice;
+}
+function useShowEarpiece(): boolean {
```
```suggestion
}

function useShowEarpiece(): boolean {
```

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -59,12 +60,26 @@ export interface MediaDevice {
+  const [alwaysShowIphoneEarpice] = useSetting(alwaysShowIphoneEarpieceSetting);
+  const m = useMemo(
+    () =>
+      (navigator.userAgent.match("iPhone")?.length ?? 0) > 0 ||
```
Is there a reason we can't do this here?

```suggestion
      navigator.userAgent.includes("iPhone") ||
```

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -74,7 +89,7 @@ function useMediaDevice(
 ): MediaDevice {
   // Make sure we don't needlessly reset to a device observer without names,
   // once permissions are already given
```
This comment belongs with the `hasRequestedPermissions` line

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -249,6 +253,144 @@ export const MediaDevicesProvider: FC<Props> = ({ children }) => {
+  const showEarpiece = useShowEarpiece();
+
+  const available = useObservableEagerState(
+    useObservable(() =>
+      setOutputDevices$.pipe(
+        startWith<OutputDevice[]>([]),
+        map((devices) => {
+          const devicesMap = new Map<string, DeviceLabel>(
+            devices.map(({ id, name }) => [id, { type: "name", name }]),
+          );
+          if (showEarpiece)
+            devicesMap.set(EARPIECE_CONFIG_ID, { type: "earpiece" });
+          return devicesMap;
+        }),
+      ),
+    ),
```
It's not safe to reference `showEarpiece` here inside `useObservable` due to https://observable-hooks.js.org/guide/gotchas.html#epic-closure

Instead you could pass `showEarpiece` to the inputs array of `useObservable` like so:

```suggestion
    useObservable((showEarpiece$) =>
      combineLatest([setOutputDevices$.pipe(startWith<OutputDevice[]>([])), showEarpiece$], (devices, showEarpiece) => {
        const devicesMap = new Map<string, DeviceLabel>(
          devices.map(({ id, name }) => [id, { type: "name", name }]),
        );
        if (showEarpiece)
          devicesMap.set(EARPIECE_CONFIG_ID, { type: "earpiece" });
        return devicesMap;
      }),
      [showEarpiece],
    ),
```

Or you could even convert `useShowEarpiece` from a React hook into an Observable, if that feels nicer.

Oh, or I suppose replacing `useObservable` with `useMemo` would work too. Lots of options :)

toger5 replied:
I ended up witht he combine latest approach

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -249,6 +253,144 @@ export const MediaDevicesProvider: FC<Props> = ({ children }) => {
+      ),
+    ),
+  );
+  const earpiceDevice = useObservableEagerState(
```
```suggestion
  const earpieceDevice = useObservableEagerState(
```

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -249,6 +253,144 @@ export const MediaDevicesProvider: FC<Props> = ({ children }) => {
+    (id: string) => {
+      if (id === EARPIECE_CONFIG_ID) {
+        setAsEarpiece(true);
+        if (earpiceDevice) setPreferredId(earpiceDevice.id);
```
I feel like we should log a warning or something if there is no `earpieceDevice`

toger5 replied:
With the recent developments in how we need to solve it on ios we actually need the case that there is no earpiece device.

toger5 replied:
If bt headphones are selected the user first has to use the native picker to go back to speaker and then the earpice can be selected.
Modeling it with earpice being available in the list or not is the most elegant we can do I think .

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -249,6 +253,144 @@ export const MediaDevicesProvider: FC<Props> = ({ children }) => {
+
+  const [preferredId, setPreferredId] = useSetting(audioOutputSetting);
+
+  const selectedId = useMemo(() => {
+    if (available.size) {
+      // If the preferred device is available, use it. Or if every available
+      // device ID is falsy, the browser is probably just being paranoid about
+      // fingerprinting and we should still try using the preferred device.
+      // Worst case it is not available and the browser will gracefully fall
+      // back to some other device for us when requesting the media stream.
+      // Otherwise, select the first available device.
+      return (preferredId !== undefined && available.has(preferredId)) ||
+        (available.size === 1 && available.has(""))
+        ? preferredId
+        : available.keys().next().value;
+    }
+    return undefined;
+  }, [available, preferredId]);
```
This duplicates some code earlier in this file - could be turned into a reusable function

toger5 replied:
good idea.
I also removed all the earpiece logic from the non controlled logic branch (It will only be used on ios where we also use controlled outputs)

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -249,6 +253,144 @@ export const MediaDevicesProvider: FC<Props> = ({ children }) => {
+  }, [available, preferredId]);
+
+  useEffect(() => {
+    if (selectedId === EARPIECE_CONFIG_ID)
+      if (selectedId !== undefined)
+        window.controls.onOutputDeviceSelect?.(selectedId);
```
Why does this only happen if `selectedId === EARPIECE_CONFIG_ID`?

toger5 replied:
yes needs to be done in any case. Updated it so that it will be done with the correct id but in dependent of the `selectedId === EARPIECE_CONFIG_ID` condition

toger5 committed:
Quickfix for testing

toger5 committed:
refactor
 - `MediaDevice`->`MediaDeviceHandle`
 - use just one provider and switch inside the
 MediaDevicesProvider between: controlledAudioOutput, webViewAudioOutput
 - fix muteAllAudio

toger5 committed:
rename setOutputDevices-> setAvailableOutputDevices

toger5 committed:
fix mute all audio via controls

toger5 committed:
review cleanup

toger5 committed:
Merge branch 'livekit' into robin/audio-output-controls

toger5 committed:
test for mute all audio

toger5 committed:
add change audio button with callback on ios

toger5 committed:
Hide the input list on both, android+ios.
This will now be decided based on controlled media devices.

toger5 committed:
Add flags to optimize EC device handling

toger5 committed:
`isBluetooth` -> `isExternalHeadset`

toger5 committed:
bump livekit client (the current version has an issue on safari)

toger5 committed:
better logging

toger5 committed:
fix no audio thinko.

toger5 committed:
actually test the impl

toger5 committed:
Allow some controls to be set before the call view is loaded.

toger5 committed:
inform ios about earpice mode

toger5 committed:
smaller diff

toger5 committed:
logger instead of native window picker

toger5 committed:
remove the whole button on click logic

toger5 committed:
change label on button

toger5 committed:
use normal button

toger5 committed:
ITS A FORM NOOOOO

toger5 committed:
make the button prettty again

toger5 committed:
back to non translated label

toger5 committed:
fix start with for output devices.

toger5 committed:
Back to translated button

toger5 committed:
rename everything to `controlledMediaDevices` to make it consistent with the actual name in the url parameters (controlledOutput is now only used for the hook that provides the output handle)

toger5 committed:
earpice -> earpiece

robintown commented on docs/controls.md:
```diff
@@ -1,7 +1,21 @@
+
+These functions must be used in conjunction with the `controlledMediaDevices` URL parameter in order to have any effect.
+
+- `controls.setAvailableOutputDevices(devices: { id: string, name: string, forEarpiece?: boolean, isEarpiece?: boolean isSpeaker?: boolean, isExternalHeadset?, boolean;}[]): void` Sets the list of available audio outputs. `forEarpiece` is used on ios only.
```
```suggestion
- `controls.setAvailableOutputDevices(devices: { id: string, name: string, forEarpiece?: boolean, isEarpiece?: boolean isSpeaker?: boolean, isExternalHeadset?, boolean; }[]): void` Sets the list of available audio outputs. `forEarpiece` is used on iOS only.
```

robintown commented on docs/controls.md:
```diff
@@ -1,7 +1,21 @@
+- `controls.setAvailableOutputDevices(devices: { id: string, name: string, forEarpiece?: boolean, isEarpiece?: boolean isSpeaker?: boolean, isExternalHeadset?, boolean;}[]): void` Sets the list of available audio outputs. `forEarpiece` is used on ios only.
+  It flags the device that should be used if the user selects earpiece mode. This should be the main stereo loudspeaker of the device.
+- `controls.onOutputDeviceSelect: ((id: string) => void) | undefined` Callback called whenever the user or application selects a new audio output.
+- `controls.setOutputDevice(id: string): void` Sets the selected audio device in EC menu. This should be used if the os decides to automatically switch to bluetooth.
```
```suggestion
- `controls.setOutputDevice(id: string): void` Sets the selected audio device in Element Call's menu. This should be used if the OS decides to automatically switch to Bluetooth, for example.
```

Generally I'm trying to keep these docs in EC descriptive of how the APIs *could* be used rather than how downstream projects do happen to use them currently.

robintown commented on docs/controls.md:
```diff
@@ -1,7 +1,21 @@
+  It flags the device that should be used if the user selects earpiece mode. This should be the main stereo loudspeaker of the device.
+- `controls.onOutputDeviceSelect: ((id: string) => void) | undefined` Callback called whenever the user or application selects a new audio output.
+- `controls.setOutputDevice(id: string): void` Sets the selected audio device in EC menu. This should be used if the os decides to automatically switch to bluetooth.
+- `controls.setOutputEnabled(enabled: boolean)` Enables/disables all audio output from the application. This can be useful for temporarily pausing audio while the controlling application is switching output devices. Output is enabled by default.
```
Just curious: do we still need this? I thought I remembered you saying that disabling audio output didn't actually help, in testing - not sure if I'm recalling that correctly

toger5 replied:
Yes that is correct. We do have another possible usecase however.
Since we might want the audio mute button, there is a chance we want to enable the audio again if the user presses the hardware volume up button. Also we might want a "hardware down + lowest volume X times" to mute the audio.
This is a simple enough interface control this.
Once could still argue that this can be dropped from this pr.

robintown replied:
Makes sense to me, I suggest we leave it in and just drop the sentence "This can be useful for temporarily pausing audio while the controlling application is switching output devices" then.

robintown commented on docs/controls.md:
```diff
@@ -1,7 +1,21 @@
+- `controls.setOutputDevice(id: string): void` Sets the selected audio device in EC menu. This should be used if the os decides to automatically switch to bluetooth.
+- `controls.setOutputEnabled(enabled: boolean)` Enables/disables all audio output from the application. This can be useful for temporarily pausing audio while the controlling application is switching output devices. Output is enabled by default.
+- `showNativeOutputDevicePicker: () => void`. This callback will be code by the webview if the user presses the output button in the settings menu.
+  This button is only shown on ios. (`userAgent.includes("IPhone")`)
```
```suggestion
- `showNativeOutputDevicePicker: (() => void) | undefined`. Callback called whenever the user presses the output button in the settings menu.
  This button is only shown on iOS. (`userAgent.includes("iPhone")`)
```

To align with the other callback docs

robintown commented on src/controls.ts:
```diff
@@ -5,15 +5,49 @@ SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
+ */
 export const setPipEnabled$ = new Subject<boolean>();
+// BehaviorSubject since the client might set this before we have subscribed (GroupCallView still in "loading" state)
+// We want the that has been set during loading to be be available immediately once loaded.
```
```suggestion
// We want the devices that have been set during loading to be available immediately once loaded.
```

robintown commented on src/controls.ts:
```diff
@@ -5,15 +5,49 @@ SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
+  [],
+);
+// BehaviorSubject since the client might set this before we have subscribed (GroupCallView still in "loading" state)
+// We want the that has been set during loading to be be available immediately once loaded.
```
```suggestion
// We want the device that has been set during loading to be available immediately once loaded.
```

robintown commented on src/controls.ts:
```diff
@@ -5,15 +5,49 @@ SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
 export const setPipEnabled$ = new Subject<boolean>();
+// BehaviorSubject since the client might set this before we have subscribed (GroupCallView still in "loading" state)
+// We want the that has been set during loading to be be available immediately once loaded.
+export const setAvailableOutputDevices$ = new BehaviorSubject<OutputDevice[]>(
```
As long as these are BehaviorSubjects (representing stateful values rather than events), I think it makes sense to name these like values:

```suggestion
export const availableOutputDevices$ = new BehaviorSubject<OutputDevice[]>(
```

robintown commented on src/controls.ts:
```diff
@@ -5,15 +5,49 @@ SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
+);
+// BehaviorSubject since the client might set this before we have subscribed (GroupCallView still in "loading" state)
+// We want the that has been set during loading to be be available immediately once loaded.
+export const setOutputDevice$ = new BehaviorSubject<string | undefined>(
```
```suggestion
export const outputDevice$ = new BehaviorSubject<string | undefined>(
```

robintown commented on src/controls.ts:
```diff
@@ -5,15 +5,49 @@ SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
+ *
+ * This should also be used to display a darkened overlay screen letting the user know that audio is muted.
+ */
+export const setOutputEnabled$ = new Subject<boolean>();
```
This comment seems inaccurate - isn't this implemented in the MuteAllAudioModel thing?

robintown commented on src/controls.ts:
```diff
@@ -27,4 +61,17 @@ window.controls = {
+      throw new Error(
+        "Output controls are disabled. No setOutputEnabled$ observer",
+      );
+    setOutputEnabled$.next(!enabled);
```
```suggestion
    setOutputEnabled$.next(enabled);
```

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -59,24 +63,73 @@ export interface MediaDevice {
+ *   `setAvailableOutputDevices$.includes((d)=>d.forEarpiece)`
+ */
+export const iosDeviceMenu$ = alwaysShowIphoneEarpieceSetting.value$.pipe(
+  startWith(
+    alwaysShowIphoneEarpieceSetting.getValue() ||
+      navigator.userAgent.includes("iPhone"),
+  ),
```
This `startWith` should be redundant since `Setting["value$"]` refers to a behavior-like observable (an observable that emits immediately upon subscription)

toger5 replied:
I got tricked by it being stored in an Observable. Which I now notice is just a trick to hide the `.next` "write" api and making this a readonly behaviour subject basically.
But since it references to the BahaviourSubject its still inheriting the "emit on subscription" property.

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -59,24 +63,73 @@ export interface MediaDevice {
+/**
+ * Hook to get access to a mediaDevice handle for a kind. This allows to list
+ * the available devices, read and set the selected device.
+ * @param kind audio input, output or video output.
+ * @param setting The setting this handles selection should be synced with.
```
```suggestion
 * @param kind Audio input, output or video output.
 * @param setting The setting this handle's selection should be synced with.
```

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -273,6 +317,73 @@ export const MediaDevicesProvider: FC<Props> = ({ children }) => {
+        }),
+      );
+
+      return combineLatest([outputDeviceData$, iosDeviceMenu$]).pipe(
+        map(([{ devicesMap, deviceForEarpiece }, iosShowEarpiece]) => {
```
```suggestion
      return combineLatest([outputDeviceData$, iosDeviceMenu$], ({ devicesMap, deviceForEarpiece }, iosShowEarpiece) => {
```

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -273,6 +317,73 @@ export const MediaDevicesProvider: FC<Props> = ({ children }) => {
+    }),
+  );
+  const [preferredId, setPreferredId] = useSetting(audioOutputSetting);
+  useEffect(() => {
+    setOutputDevice$.subscribe((id) => {
+      if (id) setPreferredId(id);
+    });
+  }, [setPreferredId]);
```
This effect needs a clean-up function which unsubscribes

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -273,6 +317,73 @@ export const MediaDevicesProvider: FC<Props> = ({ children }) => {
+  const [asEarpiece, setAsEarpiece] = useState(false);
+
+  useEffect(() => {
+    // In earpiece mode we just sent the EARPIECE_CONFIG_ID to the native code
+    // This only happens on ios where we use the native picker.
+    // So this only is needed so that ios can know if the proximity sensor should be used or not.
+    if (selectedId) window.controls.onOutputDeviceSelect?.(selectedId);
+    setAsEarpiece(selectedId === EARPIECE_CONFIG_ID);
```
I'm finding these comments hard to follow since they're mixing a description of what EC does with a description of what downstream applications do. Is something like this still accurate?

```suggestion
    // Let the hosting application know which output device has been selected.
    // This information is probably only of interest if the earpiece mode has been
    // selected - for example, Element X iOS listens to this to determine whether it
    // should enable the proximity sensor.
    if (selectedId) window.controls.onOutputDeviceSelect?.(selectedId);
    setAsEarpiece(selectedId === EARPIECE_CONFIG_ID);
```

toger5 replied:
very accurate and undestandable to me!

robintown commented on src/settings/SettingsModal.tsx:
```diff
@@ -101,20 +104,36 @@ export const SettingsModal: FC<Props> = ({
   const [showDeveloperSettingsTab] = useSetting(developerMode);
 
   const { available: isRageshakeAvailable } = useSubmitRageshake();
+  // If we are on ios we will show a button to open the native picker.
```
```suggestion
  // If we are on iOS we will show a button to open the native audio device picker.
```

robintown commented on src/settings/SettingsModal.tsx:
```diff
@@ -101,20 +104,36 @@ export const SettingsModal: FC<Props> = ({
   const { available: isRageshakeAvailable } = useSubmitRageshake();
+  // If we are on ios we will show a button to open the native picker.
+  const iosDeviceMenu = useObservableEagerState(iosDeviceMenu$);
+  // In controlled devices we will not show the input section
```
How come? Wouldn't the `iosDeviceMenu` condition be more appropriate here?

toger5 replied:
There are 3 cases:
 - normal (input:yes + output:yes)
 - ios (input:no, output: menu button + earpice picker vs speaker in output)
 - android (input: no ouput:yes)

 we model them with the two bools iosDeviceMenu, controlledMediaDevices. where one is responsible for the input: yes/no and the other on for output: default/custom button ios thing

robintown replied:
Well, `controlledMediaDevices` relates to the output devices rather than the input devices, right? So something seems off here.

> - android (input: no ouput:yes)

Why do we want hide the input devices on Android, out of interest?

toger5 replied:
The controlled media devices bool value is responsible to control whether the input is shown. If it is true we do not show the input if it is false we do show the input selection.

So yes in controller media devices input is never shown and that is how/why they define the input visibilty

toger5 replied:
On Android we hide the input because it also uses controlled media devices and the os does not expose input selection. It combines a mic and a speaker into one media device. So we change both at once and output is easier to understand. (Main speaker vs bluetooth device in input is confusing. But in output it makes a lot of sense and that the input is switched at the same time is hopefully expected and hence not too confusing for the user.)

robintown replied:
I think I get it now: for the use cases in which `mediaDeviceControls` would be active, we actually expect both the audio input devices and audio output devices to reflect the very same device list? If so, I would suggest renaming the URL parameter to `audioDeviceControls` and likewise changing the `controls` API names & docs to reference "audio input/output devices" (or maybe just "audio devices") as a unified concept, rather than specifically "audio output devices".

I also suggest you add a code comment with the explanation you just gave for why the devices are easier to understand as output devices, in the UI - I found that helpful.

toger5 replied:
The rename also makes sense but I dont see to much value in the output vs media choice. Media is very vague so it could also contain input (and even video tbh) so I think it is acceptable.
We also have rust and mobile code that already uses the naming: `controlledMediaDevices`.
I also think `controlled` is a good name for the url parameter since that explains the best what is actually happing, that the devices are controlled. `mediaDeviceControls` = `false` could be read as: There wont be any ui to control the media. (which is not true. The media devices will just not be controlled from the outside but will interact with the webview directly)

robintown replied:
Ah right, `controlledAudioDevices` was what I meant to suggest.

> Media is very vague so it could also contain input (and even video tbh)

That last bit, that it could apply to video devices, is how I see `controlledAudioDevices` could be an improvement. What do you think / how set-in-stone is the Rust and mobile code?

With the other names, I kind of just find them misleading if they only mention outputs. IMO we should make their usage expectations (that input + output are coupled) be reflected in their names while we have the chance:

```typescript
setAvailableAudioDevices(devices: AudioDevice[]): void;
setAudioDevice(id: string): void;
onAudioDeviceSelect?: (id: string) => void;
setAudioEnabled(enabled: boolean): void;
showNativeAudioDevicePicker?: () => void;
```

robintown commented on src/state/MuteAllAudioModel.ts:
```diff
@@ -0,0 +1,22 @@
+/**
+ * This can transition into sth more complete: `GroupCallViewModel.ts`
+ */
+export const muteAllAudio$ = combineLatest([
+  setOutputEnabled$,
+  muteAllAudioSetting.value$,
+]).pipe(
+  startWith([true, muteAllAudioSetting.getValue()]),
+  map(([outputEnabled, settingsMute]) => !outputEnabled || settingsMute),
+);
```
```suggestion
export const muteAllAudio$ = combineLatest([
  setOutputEnabled$.pipe(startWith(true)),
  muteAllAudioSetting.value$,
], (outputEnabled, settingsMute) => !outputEnabled || settingsMute);
```

Since the setting is already behavior-like

robintown commented on src/state/MuteAllAudioModel.test.ts:
```diff
@@ -0,0 +1,36 @@
+import { muteAllAudio as muteAllAudioSetting } from "../settings/settings";
+import { muteAllAudio$ } from "./MuteAllAudioModel";
+
+test("muteAllAudio$", () => {
```
Not a blocker, but this would be more idiomatic as a marble test

toger5 replied:
I agree. Since it s not a blocker I prefer postponing this to a later PR.

robintown reviewed (commented):
Sorry, still lots of comments, though it's mostly docs stuff

toger5 committed:
review

robintown commented on src/UrlParams.ts:
```diff
@@ -124,9 +124,15 @@ export interface UrlParams {
+   * allowing the list of output devices to be controlled by the app hosting
+   * Element Call.
+   */
+  controlledMediaDevices: boolean;
```
By the way, maybe this should instead be named `controlledOutputDevices` to reflect its behavior? I think I'm responsible for that mis-naming.

robintown replied:
Different renaming suggested at https://github.com/element-hq/element-call/pull/3270#discussion_r2101824073

toger5 committed:
Comment to explain the usage/impact of: `controlledMediaDevices`

toger5 committed:
use js-sdk branch as in livekit branch

toger5 committed:
Merge branch 'livekit' into robin/audio-output-controls

robintown commented on docs/controls.md:
```diff
@@ -1,7 +1,21 @@
+
+## Audio output devices
+
+These functions must be used in conjunction with the `controlledMediaDevices` URL parameter in order to have any effect.
```
Florian suggests this say:

> On mobile platforms (iOS, Android), web views do not reliably support selecting audio output devices such as the main speaker, earpiece, or headset. To work around this limitation, the following functions allow the hosting application (e.g., Element Web, Element X) to take control of audio device management via exposed JavaScript interfaces. These functions must be enabled by the URL parameter `controlledAudioDevices` in order to have any effect.

toger5 committed:
Deprecate old naming and introduce new words

robintown commented on docs/controls.md:
```diff
@@ -1,7 +1,21 @@
+  It flags the device that should be used if the user selects earpiece mode. This should be the main stereo loudspeaker of the device.
+- `controls.onAudioDeviceSelect: ((id: string) => void) | undefined` Callback called whenever the user or application selects a new audio output.
+- `controls.setAudioDevice(id: string): void` Sets the selected audio device in Element Call's menu. This should be used if the OS decides to automatically switch to Bluetooth, for example.
+- `controls.setAudioEnabled(enabled: boolean)` Enables/disables all audio output from the application. This can be useful for temporarily pausing audio while the controlling application is switching output devices. Output is enabled by default.
```
```suggestion
- `controls.setAudioEnabled(enabled: boolean)` Enables/disables all audio output from the application. Output is enabled by default.
```

I still suggest we drop this (since it won't be used that way after all)

toger5 replied:
okay cool.

toger5 committed:
Update docs/controls.md

Co-authored-by: Robin <robin@robin.town>

toger5 committed:
also add non deprecated audio url parameter

robintown commented on docs/url-params.md:
```diff
@@ -63,6 +63,7 @@ These parameters are relevant to both [widget](./embedded-standalone.md) and [st
 | `lang`                                                  | [BCP 47](https://www.rfc-editor.org/info/bcp47) code                                                 | No                                                                                                | No                           | The language the app should use.                                                                          ...
 | `password`                                              |                                                                                                      | No                                                                                                | No                           | E2EE password when using a shared secret. (For individual sender keys in embedded mode this is not requir ...
 | `perParticipantE2EE`                                    | `true` or `false`                                                                                    | No, defaults to `false`                                                                           | No, defaults to `false`      | Enables per participant encryption with Keys exchanged over encrypted matrix room messages.               ...
+| `controlledAudioDevices`                                | `true` or `false`                                                                                    | No, defaults to `false`                                                                           | No, defaults to `false`      | Whether the [global JS controls for audio output devices](./controls.md#audio-devices) should be enabled, ...
```
This should reference "audio devices" or "audio input/output devices" rather than "audio output devices"

robintown commented on docs/controls.md:
```diff
@@ -1,7 +1,21 @@
 - `controls.enablePip(): void` Puts the call interface into picture-in-picture mode. Throws if not in a call.
 - `controls.disablePip(): void` Takes the call interface out of picture-in-picture mode, restoring it to its natural display mode. Throws if not in a call.
+
+## Audio output devices
```
Same here

robintown commented on docs/url-params.md:
```diff
@@ -63,6 +63,7 @@ These parameters are relevant to both [widget](./embedded-standalone.md) and [st
 | `lang`                                                  | [BCP 47](https://www.rfc-editor.org/info/bcp47) code                                                 | No                                                                                                | No                           | The language the app should use.                                                                          ...
 | `password`                                              |                                                                                                      | No                                                                                                | No                           | E2EE password when using a shared secret. (For individual sender keys in embedded mode this is not requir ...
 | `perParticipantE2EE`                                    | `true` or `false`                                                                                    | No, defaults to `false`                                                                           | No, defaults to `false`      | Enables per participant encryption with Keys exchanged over encrypted matrix room messages.               ...
+| `controlledAudioDevices`                                | `true` or `false`                                                                                    | No, defaults to `false`                                                                           | No, defaults to `false`      | Whether the [global JS controls for audio output devices](./controls.md#audio-devices) should be enabled, ...
```
```suggestion
| `controlledAudioDevices`                                | `true` or `false`                                                                                    | No, defaults to `false`                                                                           | No, defaults to `false`      | Whether the [global JS controls for audio devices](./controls.md#audio-devices) should be enabled, allowing the list of audio devices to be controlled by the app hosting Element Call.                                                                                                                                                                                                     |
```

toger5 committed:
Update docs/url-params.md

Co-authored-by: Robin <robin@robin.town>

robintown committed:
Merge branch 'livekit' into robin/audio-output-controls

robintown committed:
Fix formatting
